## ArchaicSeeker2.0
ArchaicSeeker is a series of software for detecting archaic introgression sequences and reconstructing introgression history. The latest version of this series, ArchaicSeeker 2.0, has the following three notable improvements compared with the original version of this software. First, it can automatically determine the boundary of each introgressed sequence. Next, it is capable of tracing both known and unknown ancestral sources of a given introgressed sequence. Finally, it has the ability to reconstruct the introgression history with more sophisticated introgression models.

The usage of ArchaicSeeker 2.0 can be found in Manual for ArchaicSeeker 2.0.pdf.

If you use our software in your research, please cite the following papers:

Yuan K, Ni X, Liu C, Pan Y, Deng L, Zhang R, Gao Y, Ge X, Liu J, Ma X, Lou H, Wu T, Xu S. Refining models of archaic admixture in Eurasia with ArchaicSeeker 2.0. Nat Commun. 2021 Oct 29;12(1):6232. doi: 10.1038/s41467-021-26503-5. https://www.nature.com/articles/s41467-021-26503-5

and

Zhang R, Yuan K, Xu S. Detecting archaic introgression and modeling multiple-wave admixture with ArchaicSeeker 2.0. Star Protocols. 2022 June 17. doi: 10.1016/j.xpro.2022.101314. https://star-protocols.cell.com/protocols/1572

The link of this GitHub repo at Zenodo is as follows: https://doi.org/10.5281/zenodo.5526693.

#### Updateï¼š
2022.09: We add the pipeline to identify the archaic allele which can be found in the folder "ArchaicAllele". For detailed algorithms, you can refer to the methods in AS2 NC paper.

#### Correction to the protocol:
The statement in the top of page 4 (the reference allele in the input VCF files should be same as that of ancestral allele states) should be "the reference allele in the input VCF files should be same as that of reference genome".
